### PR TITLE
editor: save dirty editors when toggling auto-save on

### PR DESCRIPTION
**Pull request by vince-fugnitto**
#### What it does

The following pull-request updates the `toggleAutoSave` method to trigger a `saveAll` (for all dirty content), when the `auto-save` preference value is turned on. This behavior of saving dirty content across multiple editors is consistent with vscode and improves the user-experience.

#### How to test

1. start the application with a workspace present
2. ensure that `auto-save` is **off**
3. make dirty (unsaved) changes in multiple files
4. update the `auto-save` by enabling it in the `file` menu
5. the dirty editors should be saved

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

**Comment by lmcbout** [comment]
Tested on UBUNTU 18.04 and Chrome
- Set the File-> Auto Save to "OFF"
 - I opened  the file setting.json from the workspace
 - modify a few files with auto-save "OFF"
- Open menu, then select "Auto Save"
--> Since the file setting.json was in the editor, it puts the auto-save:"on" but did not save the preference file setting, so the other file did not get saved either.
This is a corner case here, otherwise, turning the "Auto Save" to "ON" works fine with any other files.

File: packages/editor/src/browser/editor-command.ts [hunk 1]
```diff
@@ -163,6 +166,15 @@ export class EditorCommandContribution implements CommandContribution {
     @inject(ResourceProvider)
     protected readonly resourceProvider: ResourceProvider;
 
+    @postConstruct()
+    protected init(): void {
+        this.editorPreferences.onPreferenceChanged(e => {
+            if (e.preferenceName === 'editor.autoSave' && e.newValue === 'on') {
+                this.shell.saveAll();
```
Comment (kittaakos) [inline packages/editor/src/browser/editor-command.ts:173]: Sorry for hijacking this PR, I am sure this solution works flawlessly but this is a generic ~issue~ inconveniency; almost every time when I update to `next` Theia in a downstream project, I have to resolve a cycle in the DI as we just inject services here and there.
Instead of injecting the `shell`, can we do the following instead?
```ts
this.commandService.executeCommand(CommonCommands.SAVE_ALL.id);
```
Or make the editor command contribution a frontend app contribs, and save a reference to the app, then you can get the shell.

Long story short, injecting thing here, and they usually cause a DI cycle after a Theia update. Can we make the modules less dependent on each other?

Comment (vince-fugnitto) [reply packages/editor/src/browser/editor-command.ts:173]: Sorry @kittaakos, I was unaware that using injections like I've done causes issues, even if `ApplicationShell` is from `@theia/core`. I can make the changes like you suggested, should we add a note in the coding guidelines (I was not aware about the DI cycle issues)?

Comment (kittaakos) [reply packages/editor/src/browser/editor-command.ts:173]: > Sorry @kittaakos, I was unaware that using injections like I've done causes issues, even if `ApplicationShell` is from `@theia/core`. I can make the changes like you suggested, should we add a note in the coding guidelines (I was not aware about the DI cycle issues)?

No worries at all. I think we can leave this PR as is, but it would be nice to have a strategy for the future as the number of possible DI cycles increases in downstream projects when the `Application`, `ApplicationShell`, `ShellLayoutRestorer`, etc are all customized.

Comment (vince-fugnitto) [reply packages/editor/src/browser/editor-command.ts:173]: In order to use `commandService` (for execute command) I'd need to inject it anyways.
Is there a preference between injecting the `ApplicationShell` versus `CommandService`?

Comment (kittaakos) [reply packages/editor/src/browser/editor-command.ts:173]: `CommandService` does not have injected members (fields). So it won't cause a DI cycle.

Comment (vince-fugnitto) [reply packages/editor/src/browser/editor-command.ts]: > `CommandService` does not have injected members (fields). So it won't cause a DI cycle.

Thank you for clarifying for me, I took the liberty to update the pull-request without the use of any new injections :+1:

File: packages/editor/src/browser/editor-command.ts [hunk 2]
```diff
@@ -163,6 +163,17 @@ export class EditorCommandContribution implements CommandContribution {
     @inject(ResourceProvider)
     protected readonly resourceProvider: ResourceProvider;
 
+    @postConstruct()
+    protected async init(): Promise<void> {
+        this.editorPreferences.onPreferenceChanged(e => {
+            if (e.preferenceName === 'editor.autoSave' && e.newValue === 'on') {
+                this.editorManager.all
```
Comment (akosyakov) [inline packages/editor/src/browser/editor-command.ts:170]: why do we save only editors? we should save all saveable. It won't work for custom editors and any widgets using resources.

File: packages/editor/src/browser/editor-command.ts (hunk 1 above)
Comment (akosyakov) [reply packages/editor/src/browser/editor-command.ts]: Could you revert latest changes please? Calling `save` on the shell is important.

File: packages/editor/src/browser/editor-command.ts (hunk 2 above)
Comment (vince-fugnitto) [reply packages/editor/src/browser/editor-command.ts]: > why do we save only editors? we should save all saveable. It won't work for custom editors and any widgets using resources.

@akosyakov I updated it to use the original approach of using the `shell` to `saveAll`.

**Comment by vince-fugnitto** [comment]
@lmcbout do you mind re-reviewing the pull-request?

> This is a corner case here, otherwise, turning the "Auto Save" to "ON" works fine with any other files.

I do not believe the comment you had about the `settings.json` is a bug.
We trigger a `saveAll` based on the preference changed event and if the `settings.json` is simply modified (off to on) this does not trigger an event, only explicitly saving the document will. We should not try to interpret that the user has unmodified changes for `autoSave` and trigger a save in my opinion.

I do not believe the comment you had about the `settings.json` is a bug as updating the preference value from `off` to `on` will not trigger a save (will not trigger an event). The file needs to be expli

**Comment by vince-fugnitto** [comment]
@lmcbout do you mind re-reviewing the pull-request?

> This is a corner case here, otherwise, turning the "Auto Save" to "ON" works fine with any other files.

I do not believe the comment you had about the `settings.json` is a bug.
We trigger a `saveAll` based on the preference changed event and if the `settings.json` is simply modified (off to on) this does not trigger an event, only explicitly saving the document will. We should not try to interpret that the user has unmodified changes for `autoSave` and trigger a save in my opinion.

**Comment by lmcbout** [comment]
Tested with commit 42a09db1f37493ee604106ff82bc85d5b641878c
On Ubuntu 18.04 and Chrome
If the file "setting.json" is **not open** in the editor, it works fine.  OK
If the file "settings.json" is **OPEN** in the editor and I set "AutoSave" = "ON"
Then it **does not autoSave**

AutoSaveWithSettingOpen

**Comment by vince-fugnitto** [comment]
@lmcbout the bug about the toggle command causing dirty changes to the `settings.json` is on `master` as well.

**Comment by vince-fugnitto** [comment]
I'll merge tomorrow if there are no objections :)
